### PR TITLE
Adjust Responsive Spacing Based on Layout

### LINKED ISSUE #141: Responsive layout and top-window size
**Describe the solution you'd like**
The UI should not have large unused spaces that block the users other apps, desktop etc (while not feeling cramped and have reasonable spaces between things)

The window size should be able to show all of the board layout (when selected) and all of the BCM layout (when selected) without user intervention.

It should allow the user to re-size to be smaller, if they want the window smaller and they either just want to see a few pins, or are willing to scroll around. We don't force a large window size.

### PR DESCRIPTION
**Pull request by sundaram123krishnan**
Fixes #141

### CODE REVIEW

**Comment by andrewdavidmackenzie** [comment]
The bahaviour is excellent and will help when we add other UI elements and widgets

File: src/piggui.rs [hunk 1]
```diff
@@ -37,6 +37,13 @@ pub enum Layout {
 
 impl Layout {
     const ALL: [Layout; 2] = [Layout::BoardLayout, Layout::BCMLayout];
+
+    fn get_spacing(&self) -> u16 {
+        match self {
+            Layout::BoardLayout => 470,
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:43]: I wouldn't delay this PR more with this change, but in the future maybe we can introduce constants for each dimension, then they can be added-up (to other constants) to get higher-level dimensions. Then changes in lower level widget sizes automatically change higher-up dimensions and you don't need to work them all out and edit manually.

File: src/piggui.rs [hunk 2]
```diff
@@ -110,9 +117,11 @@ fn main() -> Result<(), iced::Error> {
     }
 
     let window = window::Settings {
-        resizable: false,
-        position: window::Position::Centered,
-        size: iced::Size::new(1000.0, 900.0),
+        resizable: true,
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:120]: Same here: let window size be calculated (as a const) from the size of the views that it contains (their constants)

File: src/piggui.rs [hunk 3]
```diff
@@ -254,6 +263,12 @@ impl Gpio {
             }
         }
     }
+    fn get_dimensions_for_layout(layout: Layout) -> (f32, f32) {
+        match layout {
+            Layout::BoardLayout => (1100f32, 780f32),
+            Layout::BCMLayout => (800f32, 950f32),
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:269]: Same here, I'd introduce constants

File: src/piggui.rs [hunk 4]
```diff
@@ -274,10 +289,13 @@ impl Application for Gpio {
...
-            Command::perform(Self::load(env::args().nth(1)), |result| match result {
-                Ok(Some((filename, config))) => Message::ConfigLoaded((filename, config)),
-                _ => Message::None,
-            }),
+            Command::batch(vec![Command::perform(
+                Self::load(env::args().nth(1)),
+                |result| match result {
+                    Ok(Some((filename, config))) => Message::ConfigLoaded((filename, config)),
+                    _ => Message::None,
+                },
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:297]: Just formatting, no?

We need to get to the bottom of which of us is not using the standard rustfmt - could be me, I'll try and check. Just check this out and run rustfmt and see if it changes I guess...

File: src/piggui.rs [hunk 5]
```diff
@@ -293,6 +311,14 @@ impl Application for Gpio {
             }
             Message::LayoutChanged(layout) => {
                 self.chosen_layout = layout;
+                let (width, height) = Self::get_dimensions_for_layout(layout);
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:314]: WDYT about get_dimensions_for_layout returning a Size type?

File: src/piggui.rs [hunk 6]
```diff
@@ -431,16 +456,12 @@ impl Application for Gpio {
...
             .height(Length::Fill)
             .width(Length::Fill)
-            .padding(20)
+            .padding(10)
             .align_x(alignment::Horizontal::Center)
             .align_y(alignment::Vertical::Top)
             .into()
     }
 
-    fn scale_factor(&self) -> f64 {
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:440]: Great to see all those fixed sizes and this go!! Big improvement.

File: src/piggui.rs [hunk 7]
```diff
@@ -452,17 +473,26 @@ impl Application for Gpio {
 // Hardware Configuration Display
 fn hardware_view(hardware_description: &HardwareDescriptor) -> Element<'static, Message> {
     let hardware_info = Column::new()
-        .push(Text::new(format!("Hardware: {}", hardware_description.hardware)).size(20))
-        .push(Text::new(format!("Revision: {}", hardware_description.revision)).size(20))
-        .push(Text::new(format!("Serial: {}", hardware_description.serial)).size(20))
-        .push(Text::new(format!("Model: {}", hardware_description.model)).size(20))
+        .push(Text::new(format!(
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:476]: formatting again. The original should be under 80 columns and good I think, but I will check rustfmt

File: src/pin_layout.rs [hunk 8]
```diff
@@ -160,16 +160,20 @@ fn get_pin_widget(
...
-                        Message::PinFunctionSelected(
-                            board_pin_number,
-                            bcm_pin_number.unwrap(),
-                            Input(Some(selected_pull)),
-                        )
-                    })
-                    .placeholder("Select Pullup"),
-                )
+                Row::new()
+                    .push(led(16.0, pin_state))
```
Comment (andrewdavidmackenzie) [inline src/pin_layout.rs:164]: There's a fixed size for led (16.0) there that we could review in the future?

File: src/pin_layout.rs [hunk 9]
```diff
@@ -243,63 +251,73 @@ fn create_pin_view_side(
...
+            .width(Length::Fill)
             .placeholder("Select function"),
         );
 
         pin_option = pin_option.push(pin_options_row);
     }
 
+    let mut pin_name_column = Column::new()
+        .width(Length::Fixed(70f32))
+        .align_items(Alignment::Center);
```
Comment (andrewdavidmackenzie) [inline src/pin_layout.rs:268]: Add a constant for PIN_NAME_WIDTH?

File: src/pin_layout.rs [hunk 10]
```diff
@@ -243,63 +251,73 @@ fn create_pin_view_side(
...
         .align_items(Alignment::Center);
 
-    let mut pin_arrow = Row::new()
-        .width(Length::Fixed(60f32))
-        .align_items(Alignment::Center);
+    pin_name_column = pin_name_column.push(pin_name);
+
+    let mut pin_arrow_column = Column::new()
+        .align_items(Alignment::Center)
+        .width(Length::Fixed(30f32));
```
Comment (andrewdavidmackenzie) [inline src/pin_layout.rs:279]: Constant for PIN_ARROW_WIDTH?

File: src/pin_layout.rs [hunk 11]
```diff
@@ -243,63 +251,73 @@ fn create_pin_view_side(
...
     // Create the pin itself, with number and as a button
     let pin_button = button(
         Text::new(pin_description.board_pin_number.to_string())
-            .size(20)
             .horizontal_alignment(Horizontal::Center),
     )
-    .padding(10)
-    .width(Length::Fixed(40f32))
+    .padding(5)
+    .width(Length::Fixed(30f32))
```
Comment (andrewdavidmackenzie) [inline src/pin_layout.rs:300]: constant for PIN_BUTTON_WIDTH?

File: src/pin_layout.rs [hunk 12]
```diff
@@ -243,63 +251,73 @@ fn create_pin_view_side(
...
     .style(get_pin_color(pin_description).get_button_style())
     .on_press(Message::Activate(pin_description.board_pin_number));
 
+    pin_button_column = pin_button_column.push(pin_button);
```
Comment (andrewdavidmackenzie) [inline src/pin_layout.rs:304]: Do you need the extra column around the button?

File: src/pin_layout.rs [hunk 13]
```diff
@@ -243,63 +251,73 @@ fn create_pin_view_side(
...
     // Create the row of widgets that represent the pin, inverted order if left or right
     if is_left {
         Row::new()
             .push(pin_widget)
             .push(pin_option)
-            .push(pin_name)
-            .push(pin_arrow)
-            .push(pin_button)
-            .spacing(10)
+            .push(pin_name_column)
```
Comment (andrewdavidmackenzie) [inline src/pin_layout.rs:310]: These extra column wrappers are needed, right?

**Review by andrewdavidmackenzie — APPROVED** [review]
Added some comments, that are optional for you to include now - if you want to get the PR across the line now / I understand....they are mainly to make layout changes in the future easier, by having constants and then adding them up to get constants for containers...

A couple of other questions about whether extra wrapper widgets are needed...

And we need to converge on a rustfmt format, to avoid formatting changes getting into PRs....

File: src/piggui.rs (hunk 4 above)
Comment (sundaram123krishnan) [reply src/piggui.rs:297]: yes, that's only formatting

File: src/pin_layout.rs (hunk 8 above)
Comment (sundaram123krishnan) [reply src/pin_layout.rs:164]: sure

File: src/pin_layout.rs (hunk 12 above)
Comment (sundaram123krishnan) [reply src/pin_layout.rs:304]: yaa, otherwise the layout gets disurbed

**Pull request merged**